biz.aQute.repository/src/aQute/bnd/repository/maven/provider/IndexFile.java

```diff
@@ -13,403 +11,527 @@
-			}
-			IO.rename(tmp, index);
-		} finally {
-			lock.writeLock()
-				.unlock();
+	Archive find(String bsn, Version version) throws Exception {
+
+		//
+		// First try if this is a BSN, not a GAV
+		//
+
+		String[] parts = bsn.split(":");
+		if (parts.length == 1) {
+			ResourceInfo info = getBridge().getInfo(bsn, version);
+			if (info == null)
+				return null;
+
+			Resource resource = info.getResource();
+			return getArchive(resource);
 		}
 
-		lastModified = indexFile.lastModified();
+		//
+		// Handle the GAV. Notice that the GAV can also contain extension +
+		// classifier
+		//
+
+		Archive tmp = Archive.valueOf(bsn + ":" + version);
+		if (tmp == null)
+			return null;
+
+		return archives.keySet()
+			.stream()
+			.filter(archive -> archive.revision.program.equals(tmp.revision.program)
+				&& archive.revision.version.getOSGiVersion()
+					.equals(version)
+				&& tmp.classifier.equals(archive.classifier))
+			.findFirst()
+			.orElse(null);
 	}
 
-	public void save() throws Exception {
-		saveIndexFile();
+	/*
+	 * Answer the versions of a bsn If the bsn is a GA then we get the versions
+	 * from the archives we have in our index. Otherwise, we use the repository
+	 * bridge
+	 */
+
+	SortedSet<Version> versions(String bsn) throws Exception {
+
+		if (bsn.indexOf(':') > 0) {
+
+			//
+			// Is a GAV
+			//
+
+			String[] parts = bsn.split(":");
+
+			if (parts.length >= 2) {
+				String classifier = parts.length > 2 ? parts[parts.length - 1] : "";
+
+				Program p = Program.valueOf(parts[0], parts[1]);
+				if (p != null) {
+					Set<Version> collect = archives.keySet()
+						.stream()
+						.filter(archive -> archive.revision.program.equals(p) && archive.classifier.equals(classifier))
+						.map(archive -> archive.revision.version.getOSGiVersion())
+						.collect(Collectors.toSet());
+					return new TreeSet<>(collect);
+				}
+			}
+		}
+
+		//
+		// Non GAVs are left to the bridge
+		//
+
+		return getBridge().versions(bsn);
 	}
 
-	public Collection<Archive> getArchives() {
-		return descriptors.keySet();
+	String tooltip(Object[] target) throws Exception {
+
+		String tooltip = getBridge().tooltip(target);
+		switch (target.length) {
+			case 2 :
+				ResourceInfo info = getBridge().getInfo((String) target[0], (Version) target[1]);
+
+				Archive archive = getArchive(info.getResource());
+				if (archive != null) {
+					tooltip += "Coordinates: " + archive + "\n";
+				}
+		}
+		return tooltip;
 	}
 
-	public Map<Requirement, Collection<Capability>> findProviders(Collection<? extends Requirement> requirements) {
-		return descriptors.values()
+	private Archive getArchive(Resource resource) {
+		if (resource == null)
+			return null;
+
+		Archive archive = archives.entrySet()
 			.stream()
-			.map(BundleDescriptor::getResource)
-			.collect(toResourcesRepository())
-			.findProviders(requirements);
+			.filter((Map.Entry<Archive, Resource> e) -> e.getValue() == resource)
```

Comment:
Not sure why you need the cast, I could delete it safely in Eclipse.

Reply:
documentation ..